### ISSUE TITLE
Implement specific functional metadatas for Script instances that contain custom global classes

### ISSUE THREAD

**Issue by Lazy-Rabbit-2001** [godotengine/godot-proposals#11238]
Discussed in https://github.com/godotengine/godot-proposals/discussions/11390 (also a vote)

### Describe the project you are working on

Script improvement + Addons

### Describe the problem or limitation you are having in your project

Extension of https://github.com/godotengine/godot-proposals/issues/1047
During the practice of Godot, I (together with some users) have encountered these inconveniences:
1. Some custom global nodes/resources are always visible in create dialogue, which are, in practice, just for private use and polluting the dialog.
2. Some custom global nodes/resources that should be visible displays with suffix, which would be a bit incoherent with other items in the create dialogue. Sometimes we want the suffix to be something else that contains some information.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

They all have one solution that is flexible and optional: metadata.
Since a metadata is allowed to be manually added to a script in the latter's inspector, this would be solved in few fingers.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Adding some metadata - `object_show_in_dialog` and `script_suffix` respectively.
* `object_show_in_dialog`: The value should be a `bool`. When it is `true` or the metadata is not set, the node/resource will be shown in the create dialog; otherwise, the node/resource will be hidden from it.
* `script_suffix`: The value can be anything, but better be a `StringName` or `String`. This allows a user to replace the suffix of a node/resource in create dialog to a custom one:
```
Node
|- MyCustomNode (test.gd)
```
to
```
Node
|- MyCustomNode (custom suffix)
```
Also, if the suffix is an empty string, there will be no suffix shown.
This is useful for adding a pseudo namespace for each nodes, or some short notes stick to the class name.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Since the create dialog and the scene tree dock is not exposed to the engine, it is impossible.

### Is there a reason why this should be core and not an add-on in the asset library?

Same as "If this enhancement will not be used often, can it be worked around with a few lines of script?"

### Progress

- [x] [`script_suffix`](https://github.com/godotengine/godot/pull/99748)
- [x] `object_show_in_dialog`

**Comment by KoBeWi** [comment]
Related to #1047 and #4493 (for hiding the types)
[Hacking the create dialog with global settings](https://github.com/godotengine/godot/pull/100349) sounds more like a workaround.

Also not sure about having a setting to remove prefixes. I mean, wanting to change or remove them makes sense, but a global setting that handles individual scripts seems weird.

**Comment by KoBeWi** [comment]
Sorry, I meant suffix.

**Comment by KoBeWi** [comment]
The problem with this is that if we eventually get a proper way to hide classes from CreateDialog (via annotation or special keyword or whatever), the settings will become some subpar legacy thing.
The suffix could also be customizable with an annotation, or maybe we can add script metadata like you suggested. The settings are nothing more than a temporary workaround.

While I'm personally rather against the current implementation, it's not a final decision. It would be nice if more contributors/users could voice their opinion. So far the proposal has mixed reaction.

**Comment by HolonProduction** [comment]
I'm very skeptical about the suffix part in general. This is information that the dialogue gives about the kind of thing I am going to add. I want this information from the dialogue and I don't want e.g. plugin authors to masquerade their custom script types as native types in the dialogue. Those kinds of types have different properties, so for me this is important information, not a canvas to add custom information.

If the suffix is too disruptive when navigating the dialogue, it could be reduced to an icon on the right side which has the script path as hover hint. But making it overridable has the risk of adding lots of noise when using a combination of different plugins which utilize it differently, and it would remove information from the dialogue, which might be important to some users (me 😢).

**Comment by dalexeev** [comment]
I agree with HolonProduction. In my opinion, it was never intended to be a suffix, just an indicator to differentiate global script classes from native ones. We could replace the script file name with an icon and add filters to the Create New Node/Resource and Search Help dialogs, see:

* #4285
* #8741

I don't think we should add the ability to customize these editor dialogs. But if for some reason we decide to do so, it would be more appropriate to do it through an editor plugin, not through script metadata or annotations.

If you want to display some additional information about the class, you can use documentation system for that:

<details>
<summary>Image</summary>

![](https://github.com/user-attachments/assets/f8d27795-36eb-4c85-8fed-08cfbe41949e)

</details>

As for hiding some custom classes from editor dialogs, there is already multiple proposals and PRs:

* #1047
* #4493
* #9298
* https://github.com/godotengine/godot/pull/91020
* https://github.com/godotengine/godot/pull/95806

**Comment by KoBeWi** [comment]
Note that you can already customize prefixes when opening custom CreateDialog (https://github.com/godotengine/godot/pull/100135). If we redesign the prefixes, this functionality might need to be removed or reworked (which can realistically happen only before 4.4 stable).

**Comment by dalexeev** [comment]
> Note that you can already customize prefixes when opening custom CreateDialog ([godotengine/godot#100135](https://github.com/godotengine/godot/pull/100135)).

In my opinion, this is excessive functionality and overengineering.

**Comment by HolonProduction** [comment]
> If we redesign the prefixes, this functionality might need to be removed or reworked (which can realistically happen only until 4.4 stable).

IMO this should have been marked experimental in the first place, and we should mark it as such before 4.4 to keep our options open (if not the whole method then at least the suffix parameter, which I'd honestly remove completly). If we keep exposing editor stuff with options that are specific to the current UI and provide stability guarantees for it, improving on the editor will become very difficult and slow.

Personally I think we should only provide stability guarantees on any editor stuff between patch versions (still keeping compat on best effort basis, but EditorInterface compatibility shouldn't hinder UX improvements IMO).
